codersdk/workspacebuilds.go

```diff
@@ -50,7 +50,7 @@ type WorkspaceBuild struct {
 	InitiatorID        uuid.UUID           `json:"initiator_id"`
 	InitiatorUsername  string              `json:"initiator_name"`
 	Job                ProvisionerJob      `json:"job"`
-	Deadline           time.Time           `json:"deadline"`
+	Deadline           NullTime            `json:"deadline,omitempty"`
```

Comment:
Based on the comment in the attached issue, it's unclear what the use-case for knowing the time is null is that a pointer wouldn't solve.

Is this related to the TypeScript generation?

Reply:
I think this is a cleaner approach from a usability perspective in Go code. That is to say, the benefit is not in the frontend talking to the backend, it's using these values in the backend.

IMO it'd be nice for us to consistently use `NullTime` over `*time.Time` in the API. Things like not being able to take a pointer of a function call `&time.Now()` are not an issue when you can write `codersdk.NewNullTime(time.Now(), true)`. Likewise when using the type, there's no need for the `if t != nil` check.

There are work-arounds for making pointer use a bit more convenient, like using `coderd/util/ptr`, but it's an extra step to using them vs accessing the concrete type directly.

One last point in favor of `NullTime`, it's safer to use. There's no risk of doing a nil pointer dereference.

Reply:
FWIW working with pointers makes me sad; I much prefer the `NullXXX` approach.
Unless we start going down the lines of having a generic `Optional[T]` type?

Reply:
I'm hesitant to stray from such strong language primitives. Do we imagine having a `NullXXX` for every type that could be nullable? This seems odd to me because embedded structs would likely be pointers with `omitempty` rather than `Null<StructName>`. Are we promoting this *only* for built-in types?

Reply:
I understand why you may be hesitant, but I don't think it will be a problem. If we strive to make the zero value useful in most cases, the need for nullable values will be pretty low. We can also look at `codersdk` (today), there we only have a couple of primitives (int, string) and time.

```
      7 *int64
      5 *time.Time
      3 *string
      2 *uuid.UUID
```

The `uuid` package already has `uuid.NullUUID`, so it would make sense to use that one instead. In fact, the occurrence of `uuid.NullUUID` is much larger in our codebase than that of `*uuid.UUID` (30 vs 3), yet we have both.

I also foresee that the need for a nullable struct will be very rare (we have none in `codersdk`, currently), and in those cases it would likely be a custom type for which we can simply implement a custom json marshaler/unmarshaler. It's unfortunate Go never got around to adding `omitzero`/`omitempty` to the `json` package, because many times that would solve the use-case for using pointer-structs (i.e. often the reason is not to detect omission, it's to create a specific type of output).